gt/main.c

```diff
@@ -81,39 +155,107 @@ gt_proc(void *arg)
 			continue;
 
 		for (i = 0; i < num_rx; i++) {
+			int ret;
+			uint8_t priority;
+			struct rte_mbuf *m = rx_bufs[i];
+			struct gt_packet_fields fields;
+			struct ggu_policy policy;
+			struct ipip_tunnel_info tunnel;
+			struct ether_hdr *new_eth;
+
 			/*
-			 * TODO Decapsulate the packets.
+			 * Decapsulate the packets.
 			 *
 			 * Only request packets and priority packets
 			 * with capabilities about to expire go through a
 			 * policy decision.
 			 *
 			 * Other packets will be fowarded directly.
 			 */
+			ret = decapsulate(m, &priority, &tunnel);
+			if (ret < 0) {
+				rte_pktmbuf_free(m);
+				continue;
+			}
+
+			fields.outer_ip_ver = tunnel.flow.proto;
+			if (fields.outer_ip_ver == ETHER_TYPE_IPv4)
+				fields.gatekeeper_server_ip.v4 =
+					tunnel.flow.f.v4.src;
+			else if (fields.outer_ip_ver == ETHER_TYPE_IPv6)
+				rte_memcpy(fields.gatekeeper_server_ip.v6,
+					tunnel.flow.f.v6.src,
+					sizeof(fields.gatekeeper_server_ip.
+					v6));
```

Comment:
Another solution is to combine `decapsulate()` and `extract_packet_fields()` to create `gt_parse_incoming_pkt()` which would assign the pointers to all expected headers, and fail otherwise. Once all the packet is parsed is rest of the code should be streamlined.